README.md
# influxdb-relay

[![License][license-img]][license-href]

1. [Overview](#overview)
2. [Description](#description)
3. [Requirements](#requirements)
4. [Setup](#setup)
5. [Usage](#usage)
6. [Limitations](#limitations)
7. [Development](#development)
8. [Miscellaneous](#miscellaneous)

## Overview

Maintained fork of [influxdb-relay][overview-href] originally developed by
InfluxData. Replicate InfluxDB data for high availability.

## Description

This project adds a basic high availability layer to InfluxDB. With the right
architecture and disaster recovery processes, this achieves a highly available
setup.

## Tested on

- [Go](https://golang.org/doc/install) 1.7.4 to 1.11.2
- [InfluxDB](https://docs.influxdata.com/influxdb/v1.6/introduction/installation/) 1.6.4 to 1.7

Other versions will probably work but are untested.

## Setup

### Go

Download the daemon into your `$GOPATH` and install it in `/sur/sbin`.

```sh
go get -u github.com/vente-privee/influxdb-relay
cp ${GOPATH}/bin/influxdb-relay /usr/bin/influxdb-relay
chmod 755 /usr/bin/influxdb-relay
```

Create the configuration file in `/etc/influxdb-relay`.

```sh
mkdir -p /etc/influxdb-relay
cp ${GOPATH}/src/github.com/vente-privee/influxdb-relay/examples/sample.conf \
   /etc/influxdb-relay/influxdb-relay.conf
```

### Docker

Build your own image.

```sh
git clone git@github.com:vente-privee/influxdb-relay
cd influxdb-relay
docker build \
       --file Dockerfile \
       --rm \
       --tag influxdb-relay:latest \
       .
docker run \
       --volume /path/to/influxdb-relay.conf:/etc/influxdb-relay/influxdb-relay.conf
       --rm
       influxdb-relay:latest
```

or

Docker pull our image.

```sh
docker pull vptech/influxdb-relay:latest
docker run \
       --volume /path/to/influxdb-relay.conf:/etc/influxdb-relay/influxdb-relay.conf
       --rm
       vptech/influxdb-relay:latest
```

## Usage

You can find more documentation in [docs](docs) folder.

* [Architecture](docs/architecture.md)
* [Buffering](docs/buffering.md)
* [Caveats](docs/caveats.md)
* [Recovery](docs/recovery.md)
* [Filters](docs/filters.md)
* [Sharding](docs/sharding.md)

You can find some configurations in [examples](examples) folder.

* [sample.conf](examples/sample.conf)
* [sample_buffered.conf](examples/sample_buffered.conf)
* [kapacitor.conf](examples/kapacitor.conf)

### Configuration

```toml
[[http]]
# Name of the HTTP server, used for display purposes only.
name = "example-http"

# TCP address to bind to, for HTTP server.
bind-addr = "127.0.0.1:9096"

# Timeout for /health route
# After this time, the host may be considered down
health-timeout-ms = 10000

# Request limiting (Applied to all backend)
rate-limit = 5
burst-limit = 10

# Ping response code, default is 204
default-ping-response = 200

# Enable HTTPS requests.
ssl-combined-pem = "/path/to/influxdb-relay.pem"

# InfluxDB instances to use as backend for Relay
[[http.output]]
# name: name of the backend, used for display purposes only.
name="local-influxdb01"

# location: full URL of the /write endpoint of the backend
location="http://127.0.0.1:8086/"

# endpoints: Routes to use on Relay
# write: Route for standard InfluxDB request
# write_prom: Route for Prometheus request
# ping: Route for ping request
# query: Route fot querying InfluxDB backends
endpoints = {write="/write", write_prom="/api/v1/prom/write", ping="/ping", query="/query"}

# timeout: Go-parseable time duration. Fail writes if incomplete in this time.
timeout="10s"

# skip-tls-verification: skip verification for HTTPS location. WARNING: it's insecure. Don't use in production.
skip-tls-verification = false

# InfluxDB
[[http.output]]
name="local-influxdb02"
location="http://127.0.0.1:7086/"
endpoints = {write="/write", ping="/ping", query="/query"}
timeout="10s"

# Prometheus
[[http.output]]
name="local-influxdb03"
location="http://127.0.0.1:6086/"
endpoints = {write="/write", write_prom="/api/v1/prom/write", ping="/ping", query="/query"}
timeout="10s"

[[http.output]]
name="local-influxdb05"
location="http://127.0.0.1:5086/"
endpoints = {write="/write", write_prom="/api/v1/prom/write", ping="/ping", query="/query"}
timeout="10s"

[[udp]]
# Name of the UDP server, used for display purposes only.
name = "example-udp"

# UDP address to bind to.
bind-addr = "127.0.0.1:9096"

# Socket buffer size for incoming connections.
read-buffer = 0 # default

# Precision to use for timestamps
precision = "n" # Can be n, u, ms, s, m, h

# InfluxDB instance to use as backend for Relay.
[[udp.output]]
# name: name of the backend, used for display purposes only.
name="local1"

# location: host and port of backend.
location="127.0.0.1:8089"

# mtu: maximum output payload size
mtu=512

[[udp.output]]
name="local2"
location="127.0.0.1:7089"
mtu=1024
```

InfluxDB Relay is able to forward from a variety of input sources, including:

* `influxdb`
* `prometheus`

### Administrative tasks

#### /admin endpoint

Whereas data manipulation relies on the `/write` endpoint, some other features
such as database or user management are based on the `/query` endpoint. As
InfluxDB Relay does not send back a response body to the client(s), we are not
able to forward all of the features this endpoint provides. Still, we decided
to expose it through the `/admin` route.

It is now possible to query the `/admin` endpoint. Its usage is the same as the
standard `/query` Influx DB enpoint:

```
curl -X POST "http://127.0.0.1:9096/admin" --data-urlencode 'q=CREATE DATABASE some_database'
```

Errors will be logged just like regular `/write` queries. The HTTP response
bodies will not be forwarded back to the clients.

#### /health endpoint

This endpoint provides a quick way to check the state of all the backends.
It will return a JSON object detailing the status of the backends like this : 

```json
{
  "status": "problem",
  "problem": {
    "local-influxdb01": "KO. Get http://influxdb/ping: dial tcp: lookup influxdb on 0.0.0.0:8086: no such host"
  },
  "healthy": {
    "local-influxdb02": "OK. Time taken 3ms"
  }
}
```

If the relay encounters an error while checking a backend, this backend will be reported with the associated error in the `problems` object.
The backends wich the relay was able to communicate with will be reported in the healthy object.

The status field is a summary of the general state of the backends, the defined states are as follows:
* `healthy`: no errors were encountered
* `problem`: some backends, but no all of them, returned errors
* `critical`: every backend returned an error

### Filters

We allow tags and measurements filtering through regular expressions. Please,
take a look at [this document](docs/filters.md) for more information.

## Limitations

So far, this is compatible with Debian, RedHat, and other derivatives.

## Development

Please read carefully [CONTRIBUTING.md][contribute-href] before making a merge
request.

Clone repository into your `$GOPATH`.

```sh
mkdir -p ${GOPATH}/src/github.com/vente-privee
cd ${GOPATH}/src/github.com/vente-privee
git clone git@github.com:vente-privee/influxdb-relay
```

Enter the directory and build the daemon.

```sh
cd ${GOPATH}/src/github.com/vente-privee/influxdb-relay
go build -a -ldflags '-extldflags "-static"' -o influxdb-relay
```

## Miscellaneous

```
    ╚⊙ ⊙╝
  ╚═(███)═╝
 ╚═(███)═╝
╚═(███)═╝
 ╚═(███)═╝
  ╚═(███)═╝
   ╚═(███)═╝
```

[license-img]: https://img.shields.io/badge/license-MIT-blue.svg
[license-href]: LICENSE
[overview-href]: https://github.com/influxdata/influxdb-relay
[contribute-href]: CONTRIBUTING.md


CHANGELOG.md
Jun 28 2018 Antoine MILLET <amillet@vente-privee.com>
	From https://github.com/influxdata/influxdb-relay

Jun 29 2018 Alexandre BESLIC <abeslic@abronan.com>
	* Switch to dep, make the project buildable after go get

Jun 29 2018 Dejan FILIPOVIC <dfilipovic@vente-privee.com>
	* Add /status route

Jun 30 2018 Antoine MILLET <amillet@vente-privee.com>
	* New README structure
	* Add basic tests with golint & pylint
	* Add CHANGELOG
	* Add CONTRIBUTING guide
	* Merge https://github.com/influxdata/influxdb-relay/pull/65
	* Merge https://github.com/influxdata/influxdb-relay/pull/52
	* Merge https://github.com/influxdata/influxdb-relay/pull/59
	* Merge https://github.com/influxdata/influxdb-relay/pull/43
	* Merge https://github.com/influxdata/influxdb-relay/pull/57

Nov 15 2018 Maxime CORBIN <mcorbin@vente-privee.com>
    * Add Prometheus input support
    * Add `/admin` route to administrate underlying databases
    * Add code coverage / unit tests
    * Code refactor
    * Improve buffering feature avoiding connexions hanging
    * Improve `/ping` route
    * Improve logging
    * Add `-version` option
    
Dec 13 2018 Maxime CORBIN <mcorbin@vente-privee.com>
    * Add `/admin` endpoint that can be used to create or remove databases
    * Add `/health` endpoint that can be used to monitor the health status of every backend
    * Fixed some performance bugs & added a few more logs
    
Dec 27 2018 Clément CORNUT <ccornut@vente-privee.com>
    * Add `/admin/flush` endpoint that can be used to flush internal buffer
    * Add Rate limiting on backend
    * Add Filters for tags and measurements
    * New configuration formatting
    * Various bug fixes

relay/http_middlewares_test.go
package relay

import (
	"bytes"
	"compress/gzip"
	"io/ioutil"
	"log"
	"net/http"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/vente-privee/influxdb-relay/config"
	"time"
)

// Body is an empty body which implements io.Reader in order to create valid http.Requests
type Body struct {
	buf *bytes.Buffer
}

func (b Body) Read(p []byte) (int, error)   { return b.buf.Read(p) }
func (b *Body) Write(p []byte) (int, error) { return b.buf.Write(p) }
func (b Body) Close() error                 { return nil }

// End is a placeholder http Handler used to test middlewares
func (h *HTTP) End(w http.ResponseWriter, r *http.Request, s time.Time) {
	return
}

// EndTag is a placeholder http Handler used to test middlewares, which changes a variable
func (h *HTTP) EndTag(w http.ResponseWriter, r *http.Request, s time.Time) {
	wasInEnd = true
	return
}

func (h *HTTP) tagMiddleware(next relayHandlerFunc) relayHandlerFunc {
	return relayHandlerFunc(func(h *HTTP, w http.ResponseWriter, r *http.Request, s time.Time) {
		wasInMiddleware = true
		next(h, w, r, s)
	})
}

// ResponseWriter is a placeholder http.ResponseWriter to give to middlewares
type ResponseWriter struct {
	writeBuf *bytes.Buffer
	code     int
	header   http.Header
}

func (r *ResponseWriter) Header() http.Header         { return r.header }
func (r *ResponseWriter) Write(p []byte) (int, error) { return r.writeBuf.Write(p) }
func (r *ResponseWriter) WriteHeader(statusCode int)  { r.code = statusCode }

type Logger struct {
	buffer *bytes.Buffer
}

func (l Logger) Write(p []byte) (n int, err error) { return l.buffer.Write(p) }

var (
	emptyBody     = Body{}
	wrongGzipBody = Body{}
	gzipBody      = Body{}

	wasInMiddleware = false
	wasInEnd        = false

	logger = Logger{buffer: bytes.NewBuffer([]byte{})}

	emptyWriter = &ResponseWriter{
		writeBuf: &bytes.Buffer{},
		header:   http.Header{},
	}

	errorDbWriter = &ResponseWriter{
		writeBuf: bytes.NewBuffer([]byte("\"missing parameter: db\"")),
		header: http.Header{
			"Content-Length": []string{"23"},
			"Content-Type":   []string{"application/json"},
		},
		code: 400,
	}

	errorGzipWriter = &ResponseWriter{
		writeBuf: bytes.NewBuffer([]byte("\"unable to decode gzip body\"")),
		header: http.Header{
			"Content-Length": []string{"28"},
			"Content-Type":   []string{"application/json"},
		},
		code: 400,
	}
)

var (
	// w is a ResponseWriter that will be given to the middlewares
	w = &ResponseWriter{
		writeBuf: &bytes.Buffer{},
		header:   http.Header{},
	}
)

func init() {
	wrongGzipBody.buf = &bytes.Buffer{}
	_, err := wrongGzipBody.Write([]byte("WRONG"))
	if err != nil {
		log.Fatal(err)
	}

	gzipBody.buf = &bytes.Buffer{}
	gzipWriter := gzip.NewWriter(gzipBody.buf)
	_, err = gzipWriter.Write([]byte("GOOOOD"))
	if err != nil {
		log.Fatal(err)
	}
	if err = gzipWriter.Close(); err != nil {
		log.Fatal(err)
	}
}

func resetWriter() {
	w.writeBuf = &bytes.Buffer{}
	w.code = 0
	w.header = http.Header{}
}

func WriterTest(t *testing.T, expected *ResponseWriter, actual *ResponseWriter) {
	assert.Equal(t, expected.code, actual.code)
	expectedBuf, _ := ioutil.ReadAll(expected.writeBuf)
	actualBuf, _ := ioutil.ReadAll(actual.writeBuf)
	assert.Equal(t, string(expectedBuf), string(actualBuf))
	assert.Equal(t, expected.header, actual.header)
}

func createHTTP(t *testing.T, cfg config.HTTPConfig, verbose bool) *HTTP {
	tmp, err := NewHTTP(cfg, verbose, config.Filters{})
	if err != nil {
		t.Fatal(err)
	}
	return tmp.(*HTTP)
}

func TestLogMiddleware(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{}, true)

	h.logger = log.New(logger, "", 0)
	handler := h.logMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", emptyBody)
	if err != nil {
		t.Fatal(err)
	}
	handler(h, w, r, ti)
	buf, _ := ioutil.ReadAll(logger.buffer)
	assert.Equal(t, string(buf), "got request on: influxdb\n")
}

func TestLogMiddlewareNoLog(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{}, false)

	h.logger = log.New(logger, "", 0)
	handler := h.logMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", emptyBody)
	if err != nil {
		t.Fatal(err)
	}
	handler(h, w, r, ti)
	buf, _ := ioutil.ReadAll(logger.buffer)
	assert.Equal(t, string(buf), "")
}

func TestQueryMiddleware(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{}, false)

	handler := h.queryMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", emptyBody)
	if err != nil {
		t.Fatal(err)
	}
	url, _ := r.URL.Parse("http://influxdb:8086/write?db=mydb")
	r.URL = url
	handler(h, w, r, ti)
	assert.Equal(t, emptyWriter, w)
}

func TestQueryMiddlewareNoDB(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{}, false)

	handler := h.queryMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", emptyBody)
	if err != nil {
		t.Fatal(err)
	}
	url, _ := r.URL.Parse("http://influxdb:8086/write")
	r.URL = url
	handler(h, w, r, ti)
	WriterTest(t, errorDbWriter, w)
}

func TestQueryMiddlewareRp(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{DefaultRetentionPolicy: "patate"}, false)

	handler := h.queryMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", emptyBody)
	if err != nil {
		t.Fatal(err)
	}
	url, _ := r.URL.Parse("http://influxdb:8086/write?db=mydb")
	r.URL = url
	handler(h, w, r, ti)
	assert.Equal(t, emptyWriter, w)
	queryParams := r.URL.Query()
	assert.Equal(t, "patate", queryParams.Get("rp"))
}

func TestBodyMiddleWareNoGzip(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{}, false)

	handler := h.bodyMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", emptyBody)
	if err != nil {
		t.Fatal(err)
	}
	handler(h, w, r, ti)
	assert.Equal(t, emptyWriter, w)
}

func TestBodyMiddleWareErrorGzip(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{}, false)

	handler := h.bodyMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", wrongGzipBody)
	if err != nil {
		t.Fatal(err)
	}
	r.Header.Set("Content-Encoding", "gzip")
	handler(h, w, r, ti)
	WriterTest(t, errorGzipWriter, w)
}

func TestBodyMiddleWareGzip(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{}, false)
	handler := h.bodyMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", gzipBody)
	if err != nil {
		t.Fatal(err)
	}
	r.Header.Set("Content-Encoding", "gzip")
	handler(h, w, r, ti)
	buf, _ := ioutil.ReadAll(r.Body)
	assert.Equal(t, "GOOOOD", string(buf))
}

func TestAllMiddlewares(t *testing.T) {
	h := createHTTP(t, config.HTTPConfig{}, false)
	r, err := http.NewRequest("", "influxdb", gzipBody)
	if err != nil {
		t.Fatal(err)
	}

	middlewares = []relayMiddleware{
		(*HTTP).tagMiddleware,
	}
	chain := allMiddlewares(h, (*HTTP).EndTag)
	chain(h, w, r, ti)
	assert.Equal(t, true, wasInMiddleware)
	assert.Equal(t, true, wasInEnd)
}


main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"os/signal"

	"github.com/vente-privee/influxdb-relay/config"
	"github.com/vente-privee/influxdb-relay/relayservice"
)

const (
	relayVersion = "3.0.0"
)

var (
	usage = func() {
		fmt.Println("Please, see README for more information about InfluxDB Relay...")
		flag.PrintDefaults()
	}

	configFile  = flag.String("config", "", "Configuration file to use")
	verbose     = flag.Bool("v", false, "If set, InfluxDB Relay will log HTTP requests")
	versionFlag = flag.Bool("version", false, "Print current InfluxDB Relay version")
)

func runRelay(cfg config.Config) {
	relay, err := relayservice.New(cfg)
	if err != nil {
		log.Fatal(err)
	}

	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, os.Interrupt)

	go func() {
		<-sigChan
		relay.Stop()
	}()

	log.Println("starting relays...")
	relay.Run()
}

func main() {
	flag.Usage = usage
	flag.Parse()

	if *versionFlag {
		fmt.Println("influxdb-relay version " + relayVersion)
		return
	}

	// Configuration file is mandatory
	if *configFile == "" {
		fmt.Fprintln(os.Stderr, "Missing configuration file")
		flag.PrintDefaults()
		os.Exit(1)
	}

	// And it has to be loaded in order to continue
	cfg, err := config.LoadConfigFile(*configFile)
	if err != nil {
		log.Println("Version: " + relayVersion)
		log.Fatal(err.Error())
	}

	cfg.Verbose = *verbose
	runRelay(cfg)
}


relay/http_middlewares.go
package relay

import (
	"compress/gzip"
	"net/http"
	"time"
)

func allMiddlewares(h *HTTP, handlerFunc relayHandlerFunc) relayHandlerFunc {
	var res = handlerFunc
	for _, middleware := range middlewares {
		res = middleware(h, res)
	}

	return res
}

func (h *HTTP) logMiddleWare(next relayHandlerFunc) relayHandlerFunc {
	return relayHandlerFunc(func(h *HTTP, w http.ResponseWriter, r *http.Request, start time.Time) {
		if h.log {
			h.logger.Println("got request on: " + r.URL.Path)
		}
		next(h, w, r, start)
	})
}

func (h *HTTP) bodyMiddleWare(next relayHandlerFunc) relayHandlerFunc {
	return relayHandlerFunc(func(h *HTTP, w http.ResponseWriter, r *http.Request, start time.Time) {
		var body = r.Body

		if r.Header.Get("Content-Encoding") == "gzip" {
			b, err := gzip.NewReader(r.Body)
			if err != nil {
				jsonResponse(w, response{http.StatusBadRequest, "unable to decode gzip body"})
				return
			}
			defer b.Close()
			body = b
		}

		r.Body = body
		next(h, w, r, start)
	})
}

func (h *HTTP) queryMiddleWare(next relayHandlerFunc) relayHandlerFunc {
	return relayHandlerFunc(func(h *HTTP, w http.ResponseWriter, r *http.Request, start time.Time) {
		queryParams := r.URL.Query()

		if queryParams.Get("db") == "" && (r.URL.Path == "/write" || r.URL.Path == "/api/v1/prom/write") {
			jsonResponse(w, response{http.StatusBadRequest, "missing parameter: db"})
			return
		}

		if queryParams.Get("rp") == "" && h.rp != "" {
			queryParams.Set("rp", h.rp)
		}

		r.URL.RawQuery = queryParams.Encode()
		next(h, w, r, start)
	})

}

func (h *HTTP) rateMiddleware(next relayHandlerFunc) relayHandlerFunc {
	return relayHandlerFunc(func(h *HTTP, w http.ResponseWriter, r *http.Request, start time.Time) {
		if h.rateLimiter != nil && !h.rateLimiter.Allow() {
			jsonResponse(w, response{http.StatusTooManyRequests, http.StatusText(http.StatusTooManyRequests)})
			return
		}

		next(h, w, r, start)
	})
}


relay/http.go
package relay

import (
	"bytes"
	"crypto/tls"
	"encoding/json"
	"errors"
	"fmt"
	"golang.org/x/time/rate"
	"io/ioutil"
	"log"
	"net"
	"net/http"
	"os"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/influxdata/influxdb/models"
	"github.com/vente-privee/influxdb-relay/config"
)

// HTTP is a relay for HTTP influxdb writes
type HTTP struct {
	addr   string
	name   string
	schema string

	cert string
	rp   string

	pingResponseCode    int
	pingResponseHeaders map[string]string

	closing int64
	l       net.Listener

	backends []*httpBackend

	start  time.Time
	log    bool
	logger *log.Logger

	rateLimiter *rate.Limiter

	healthTimeout time.Duration
}

type relayHandlerFunc func(h *HTTP, w http.ResponseWriter, r *http.Request, start time.Time)
type relayMiddleware func(h *HTTP, handlerFunc relayHandlerFunc) relayHandlerFunc

// Default HTTP settings and a few constants
const (
	DefaultHTTPPingResponse = http.StatusNoContent
	DefaultHTTPTimeout      = 10 * time.Second
	DefaultMaxDelayInterval = 10 * time.Second
	DefaultBatchSizeKB      = 512

	KB = 1024
	MB = 1024 * KB
)

var (
	handlers = map[string]relayHandlerFunc{
		"/write":             (*HTTP).handleStandard,
		"/api/v1/prom/write": (*HTTP).handleProm,
		"/ping":              (*HTTP).handlePing,
		"/status":            (*HTTP).handleStatus,
		"/admin":             (*HTTP).handleAdmin,
		"/admin/flush":				(*HTTP).handleFlush,
		"/health":            (*HTTP).handleHealth,
	}

	middlewares = []relayMiddleware{
		(*HTTP).bodyMiddleWare,
		(*HTTP).queryMiddleWare,
		(*HTTP).logMiddleWare,
		(*HTTP).rateMiddleware,
	}
)

// NewHTTP creates a new HTTP relay
// This relay will most likely be tied to a RelayService
// and manage a set of HTTPBackends
func NewHTTP(cfg config.HTTPConfig, verbose bool, fs config.Filters) (Relay, error) {
	h := new(HTTP)

	h.addr = cfg.Addr
	h.name = cfg.Name
	h.log = verbose
	h.logger = log.New(os.Stdout, "relay: ", 0)

	h.pingResponseCode = DefaultHTTPPingResponse
	if cfg.DefaultPingResponse != 0 {
		h.pingResponseCode = cfg.DefaultPingResponse
	}

	h.pingResponseHeaders = make(map[string]string)
	h.pingResponseHeaders["X-InfluxDB-Version"] = "relay"
	if h.pingResponseCode != http.StatusNoContent {
		h.pingResponseHeaders["Content-Length"] = "0"
	}

	h.cert = cfg.SSLCombinedPem
	h.rp = cfg.DefaultRetentionPolicy

	// If a cert is specified, this means the user
	// wants to do HTTPS
	h.schema = "http"
	if h.cert != "" {
		h.schema = "https"
	}

	// For each output specified in the config, we are going to create a backend
	for i := range cfg.Outputs {
		backend, err := newHTTPBackend(&cfg.Outputs[i], fs)
		if err != nil {
			return nil, err
		}

		h.backends = append(h.backends, backend)
	}

	// If a RateLimit is specified, create a new limiter
	if cfg.RateLimit != 0 {
		if cfg.BurstLimit != 0 {
			h.rateLimiter = rate.NewLimiter(rate.Limit(cfg.RateLimit), cfg.BurstLimit)
		} else {
			h.rateLimiter = rate.NewLimiter(rate.Limit(cfg.RateLimit), 1)
		}
	}

	h.healthTimeout = time.Duration(cfg.HealthTimeout) * time.Millisecond

	return h, nil
}

// Name is the name of the HTTP relay
// a default name might be generated if it is
// not specified in the configuration file
func (h *HTTP) Name() string {
	if h.name == "" {
		return fmt.Sprintf("%s://%s", h.schema, h.addr)
	}

	return h.name
}

// Run actually launch the HTTP endpoint
func (h *HTTP) Run() error {
	var cert tls.Certificate
	l, err := net.Listen("tcp", h.addr)
	if err != nil {
		return err
	}

	// support HTTPS
	if h.cert != "" {
		cert, err = tls.LoadX509KeyPair(h.cert, h.cert)
		if err != nil {
			return err
		}

		l = tls.NewListener(l, &tls.Config{
			Certificates: []tls.Certificate{cert},
		})
	}

	h.l = l

	if h.log {
		h.logger.Printf("starting %s relay %q on %v", strings.ToUpper(h.schema), h.Name(), h.addr)
	}

	err = http.Serve(l, h)
	if atomic.LoadInt64(&h.closing) != 0 {
		return nil
	}
	return err
}

// Stop actually stops the HTTP endpoint
func (h *HTTP) Stop() error {
	atomic.StoreInt64(&h.closing, 1)
	return h.l.Close()
}

// ServeHTTP is the function that handles the different route
// The response is a JSON object describing the state of the operation
func (h *HTTP) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	// h.start = time.Now()

	if fun, ok := handlers[r.URL.Path]; ok {
		allMiddlewares(h, fun)(h, w, r, time.Now())
	} else {
		jsonResponse(w, response{http.StatusNotFound, http.StatusText(http.StatusNotFound)})
		return
	}
}

type responseData struct {
	ContentType     string
	ContentEncoding string
	StatusCode      int
	Body            []byte
}

func (rd *responseData) Write(w http.ResponseWriter) {
	if rd.ContentType != "" {
		w.Header().Set("Content-Type", rd.ContentType)
	}

	if rd.ContentEncoding != "" {
		w.Header().Set("Content-Encoding", rd.ContentEncoding)
	}

	w.Header().Set("Content-Length", strconv.Itoa(len(rd.Body)))
	w.WriteHeader(rd.StatusCode)
	w.Write(rd.Body)
}

func jsonResponse(w http.ResponseWriter, r response) {
	data, err := json.Marshal(r.body)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Content-Length", fmt.Sprint(len(data)))
	w.WriteHeader(r.code)

	_, _ = w.Write(data)
}

type poster interface {
	post([]byte, string, string, string) (*responseData, error)
	getStats() map[string]string
}

type simplePoster struct {
	client   *http.Client
	location string
}

func newSimplePoster(location string, timeout time.Duration, skipTLSVerification bool) *simplePoster {
	// Configure custom transport for http.Client
	// Used for support skip-tls-verification option
	transport := &http.Transport{
		TLSClientConfig: &tls.Config{
			InsecureSkipVerify: skipTLSVerification,
		},
	}

	return &simplePoster{
		client: &http.Client{
			Timeout:   timeout,
			Transport: transport,
		},
		location: location,
	}
}

func (s *simplePoster) getStats() map[string]string {
	v := make(map[string]string)
	v["location"] = s.location
	return v
}

func (s *simplePoster) post(buf []byte, query string, auth string, endpoint string) (*responseData, error) {
	req, err := http.NewRequest("POST", s.location+endpoint, bytes.NewReader(buf))
	if err != nil {
		return nil, err
	}

	req.URL.RawQuery = query
	req.Header.Set("Content-Type", "text/plain")
	req.Header.Set("Content-Length", strconv.Itoa(len(buf)))
	if auth != "" {
		req.Header.Set("Authorization", auth)
	}

	resp, err := s.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	data, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	return &responseData{
		ContentType:     resp.Header.Get("Content-Type"),
		ContentEncoding: resp.Header.Get("Content-Encoding"),
		StatusCode:      resp.StatusCode,
		Body:            data,
	}, nil
}

type httpBackend struct {
	poster
	name      string
	inputType config.Input
	admin     string
	endpoints config.HTTPEndpointConfig
	location  string

	tagRegexps         []*regexp.Regexp
	measurementRegexps []*regexp.Regexp
}

// validateRegexps checks if a request on this backend matches
// all the tag regular expressions for this backend
func (b *httpBackend) validateRegexps(ps models.Points) error {
  // For each point
  for _, p := range ps {
    // Check if the measurement of each point
    // matches ALL measurement regular expressions
    m := p.Name()
    for _, r := range b.measurementRegexps {
      if !r.Match(m) {
        return errors.New("bad measurement")
      }
    }

    // For each tag of each point
    for _, t := range p.Tags() {
      // Check if each tag of each point
      // matches ALL tags regular expressions
      for _, r := range b.tagRegexps {
        if !r.Match(t.Key) {
          return errors.New("bad tag")
        }
      }
    }
  }

  return nil
}

func (b *httpBackend) getRetryBuffer() *retryBuffer	{
	if p, ok := b.poster.(*retryBuffer); ok {
		return p
	}

	return nil
}

func newHTTPBackend(cfg *config.HTTPOutputConfig, fs config.Filters) (*httpBackend, error) {
	// Get default name
	if cfg.Name == "" {
		cfg.Name = cfg.Location
	}

	// Set a timeout
	timeout := DefaultHTTPTimeout
	if cfg.Timeout != "" {
		t, err := time.ParseDuration(cfg.Timeout)
		if err != nil {
			return nil, fmt.Errorf("error parsing HTTP timeout '%v'", err)
		}
		timeout = t
	}

	// Get underlying Poster instance
	var p poster = newSimplePoster(cfg.Location, timeout, cfg.SkipTLSVerification)

	// If configured, create a retryBuffer per backend.
	// This way we serialize retries against each backend.
	if cfg.BufferSizeMB > 0 {
		max := DefaultMaxDelayInterval
		if cfg.MaxDelayInterval != "" {
			m, err := time.ParseDuration(cfg.MaxDelayInterval)
			if err != nil {
				return nil, fmt.Errorf("error parsing max retry time %v", err)
			}
			max = m
		}

		batch := DefaultBatchSizeKB * KB
		if cfg.MaxBatchKB > 0 {
			batch = cfg.MaxBatchKB * KB
		}

		p = newRetryBuffer(cfg.BufferSizeMB*MB, batch, max, p)
	}

	var tagRegexps []*regexp.Regexp
	var measurementRegexps []*regexp.Regexp

	// Get regexps related to this HTTP backend
	for _, f := range fs {
		for _, e := range f.Outputs {
			if e == cfg.Name {
				if f.TagRegexp != nil {
					tagRegexps = append(tagRegexps, f.TagRegexp)
				}

				if f.MeasurementRegexp != nil {
					measurementRegexps = append(measurementRegexps, f.MeasurementRegexp)
				}
			}
		}
	}

	return &httpBackend{
		poster:             p,
		name:               cfg.Name,
		tagRegexps:         tagRegexps,
		measurementRegexps: measurementRegexps,
		endpoints: cfg.Endpoints,
		location:  cfg.Location,
	}, nil
}

// ErrBufferFull error indicates that retry buffer is full
var ErrBufferFull = errors.New("retry buffer full")

var bufPool = sync.Pool{New: func() interface{} { return new(bytes.Buffer) }}

func getBuf() *bytes.Buffer {
	if bb, ok := bufPool.Get().(*bytes.Buffer); ok {
		return bb
	}
	return new(bytes.Buffer)
}

func putBuf(b *bytes.Buffer) {
	b.Reset()
	bufPool.Put(b)
}
